engines/e_cng.c

```diff
@@ -0,0 +1,1586 @@
+{
+    char errstr[20];
+
+    BIO_snprintf(errstr, sizeof(errstr), "%lX", err);
+    ERR_add_error_data(2, "Error code=0x", errstr);
+}
+
+static CNG_CTX *cng_ctx_new(void)
+{
+    return OPENSSL_zalloc(sizeof(CNG_CTX));
+}
+
+static void cng_ctx_free(CNG_CTX *ctx)
+{
+    if (ctx == NULL)
+        return;
+    if (ctx->khandle != 0)
+        NCryptFreeObject(ctx->khandle);
+    OPENSSL_free(ctx);
+}
+
+static char *to_asc(LPCWSTR in)
+{
+    char *str;
+    int outsize;
+
+    outsize = WideCharToMultiByte(CP_UTF8, 0, in, -1, NULL, 0, NULL, NULL);
+    if (outsize == 0) {
+        int err = GetLastError();
+
+        CNGerr(CNG_F_TO_ASC,CNG_R_WIDECHARTOMULTIBYTE_ERROR);
+        cng_adderror(err);
+        return NULL;
+    }
+    str = OPENSSL_malloc(outsize);
+    if (str == NULL) {
+        CNGerr(CNG_F_TO_ASC, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    if (!WideCharToMultiByte(CP_UTF8, 0, in, -1, str, outsize, NULL, NULL)) {
+        int err = GetLastError();
+
+        OPENSSL_free(str);
+        CNGerr(CNG_F_TO_ASC,CNG_R_WIDECHARTOMULTIBYTE_ERROR);
+        cng_adderror(err);
+        return NULL;
+    }
+    return str;
+}
+
+static LPWSTR to_wide(const char *in)
+{
+    LPWSTR str;
+    int outsize;
+
+    outsize = MultiByteToWideChar(CP_UTF8, 0, in, -1, NULL, 0);
+    if (outsize == 0) {
+        int err = GetLastError();
+
+        CNGerr(CNG_F_TO_WIDE,CNG_R_MULTIBYTETOWIDECHAR_ERROR);
+        cng_adderror(err);
+        return NULL;
+    }
+    str = OPENSSL_malloc(outsize * sizeof(WCHAR));
+    if (str == NULL) {
+        CNGerr(CNG_F_TO_WIDE, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    if (!MultiByteToWideChar(CP_UTF8, 0, in, -1, str, outsize)) {
+        int err = GetLastError();
+
+        OPENSSL_free(str);
+        CNGerr(CNG_F_TO_WIDE,CNG_R_MULTIBYTETOWIDECHAR_ERROR);
+        cng_adderror(err);
+        return NULL;
+    }
+    return str;
+}
+
+static EVP_PKEY *cng_get_pkey(const ENGINE *engine, NCRYPT_KEY_HANDLE khandle)
+{
+    union {
+        BCRYPT_KEY_BLOB info;
+        BCRYPT_DH_KEY_BLOB dh;
+        BCRYPT_DSA_KEY_BLOB dsa;
+# if _WIN32_WINNT >= 0x0602      /* Win 8 */
+        BCRYPT_DSA_KEY_BLOB_V2 dsa2;
+# endif
+        BCRYPT_ECCKEY_BLOB ecc;
+        BCRYPT_RSAKEY_BLOB rsa;
+        unsigned char data[1];   /* Used to reach the data after the blobs */
+    } *exported = NULL;
+    DWORD size;
+    EVP_PKEY *res = NULL;
+
+    if (NCryptExportKey(khandle, 0, BCRYPT_PUBLIC_KEY_BLOB, NULL, NULL, 0,
+                        &size, 0) != ERROR_SUCCESS) {
+        int err = GetLastError();
+
+        CNGerr(CNG_F_CNG_GET_PKEY, CNG_R_NCRYPTEXPORTKEY_ERROR);
+        cng_adderror(err);
+        goto err;
+    }
+    if ((exported = OPENSSL_zalloc(size)) == NULL) {
+        CNGerr(CNG_F_CNG_GET_PKEY, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+    if (NCryptExportKey(khandle, 0, BCRYPT_PUBLIC_KEY_BLOB, NULL,
+                        (PBYTE)exported, size, NULL, 0) != ERROR_SUCCESS) {
```

Comment:
Passing `NULL` to `pcbResult`  at this point causes `NCryptExportKey` to fail with 0x800706F4 in my test case. Passing `&size` resolves that problem.